### ISSUE TITLE
Running tests requires PhantomJS

### ISSUE THREAD

**Issue by bhollis** [middleman/middleman#1722]
`bundle install` doesn't install PhantomJS, and `rake test` doesn't check for it and help you install it. Homebrew doesn't support installing it either, so you have to search around until you find a way to install it through `npm`. We should be making Middleman easy to contribute to, and this complicates the process.

**Comment by maxmeyer** [comment]
The problem is that we need PhaontomJS for `capybara` which itself is used in the test suite.

**Comment by tdreyno** [comment]
I thought it was bundled inside capybara?

**Comment by bhollis** [comment]
Nope, I needed to install phantomjs via npm to get it working.

**Comment by maxmeyer** [comment]
... or via package manager of your os. at is this what I do normally.

**Comment by bhollis** [comment]
Right, as I mentioned, the Homebrew package for phantomjs is broken.

**Comment by tdreyno** [comment]
If we can make capybara optional, that's great for me.

It's still really hard to setup extensions to run tests using our helpers and fixtures. I think this was intended to help that, but it's still very difficult.

**Comment by maxmeyer** [comment]
Mmmh. I think I (or my alter ego @dg-ratiodata) added `capybara` to `middleman` because it makes testing of JavaScript-apps possible. At least if one uses the pre-defined server steps - just have a look for `visit` and the like - and has PhantomJS installed. There's a ruby gem we could use https://rubygems.org/gems/phantomjs.

**Comment by maxmeyer** [comment]
I added a PR to make it possible to install phantomjs via rubygems, see #1725. @tdreyno I run both the PR CI and my local test suite in parallel. So I think, it makes sense to wait for travis to finish successfully.

**Comment by bhollis** [comment]
Which Middleman tests require JavaScript execution?

**Comment by tdreyno** [comment]
Not sure, I think a lot of them where switched to capybara's API by @maxmeyer. I think it's more about having a nice API for interacting with the DOM content, than JS, but I may be wrong.

**Comment by maxmeyer** [comment]
> Which Middleman tests require JavaScript execution?

None. But from my understand you are free to use middleman's cucumber steps in your very own test suite. See the explanation in the PR which introduced the switch to `capybara`: https://github.com/middleman/middleman/pull/1535. 

From my understanding `poltergeist` is only required for those who need JavaScript testing. Middleman's test suite does *not* require `PhantomJS`, but it is used if installed by the user for feature tests flagged with `@javascript`. Maybe we could just add documentation about how to test javascript applications with middlemans steps.

WDYT?

**Comment by maxmeyer** [comment]
@bhollis Did you have a failing test?

**Comment by tdreyno** [comment]
I think a page on testing sites with Ruby & JS would be useful. I bet the process is pretty bad right now.

**Comment by dg-ratiodata** [comment]
Good. Will send you a PR to middleman-guides. Wasn't there even another issue filed by some guy with rspec + testing v4. Maybe we could address this as well.

**Comment by bhollis** [comment]
The only `@javascript` test I could find is `middleman-core/features/javascript-testing.feature`, which simply tests that you can use PhantomJS to run a test, rather than testing anything about Middleman itself. Why does this need to be part of the test suite at all?

**Comment by maxmeyer** [comment]
> PhantomJS to run a test, rather than testing anything about Middleman itself

I would consider the `cucumber`-steps as part of the public API of `middleman`. Changing them would break other's middleman setup. So from my point of view it does make sense to have such a test included.

**Comment by bhollis** [comment]
Does this mean that an implicit dependency on PhantomJS is part of the public API of Middleman? Why can't users depend on PhantomJS themselves if they need it?

**Comment by maxmeyer** [comment]
> Does this mean that an implicit dependency on PhantomJS is part of the public API of Middleman? Why can't users depend on PhantomJS themselves if they need it?

Can we agree on this?

1. PhantomJS is only needed for the test suite of middleman not for installing and running middleman itself to run your website - given the code that exists now
2. I think we're both *against* adding `phantomjs` as dependency of middleman-core itself
3. Using the term public API is "too" strong, I shouldn't have used it, but nevertheless the cucumber steps are some kind of an API users are "free" and allowed to use.
4. If users want to test javascript apps they should be aware that installing PhantomJS needs to be installed separatedly - that would be an documentation "issue"

The test mentioned early documents that it is possible to run and test middleman apps which require javascript. If I remember correctly `Capybara` uses selenium as default javascript runner. The PR only sets phantomjs as default and requires phantomjs to be installed if a users wants to test a javascript app. But the same would be true if the default would be selenium - which cannot test headless btw.

The test suite itself only requires PhantomJS to be installed because of that single test. The question is should we keep that test and document that javascript apps are supported (and can be tested) or should we avoid it to reduce the depencies of the code base for middleman developers. Correct?

**Comment by tdreyno** [comment]
I kind of think we should remove whatever core isn't using and document how to add back in things like this for actual app testing.

**Comment by dg-ratiodata** [comment]
@tdreyno So better removing that test?

**Comment by tdreyno** [comment]
yeah, I think so.

**Comment by sandstrom** [comment]
I'm doing some issue-gardening 🌿🌷🌾 and came across this issue.

With the shift to external pipeline I'd say it's easiest to simply remove PhantomJS and skip documenting how to add this back (since people will use some other JS-framework anyway, which will have it's own idea of how to wire up JS-testing).

@dg-ratiodata @maxmeyer @bhollis Apart from the file mentioned above[1], what other stuff should be removed?

[1] https://github.com/middleman/middleman/blob/master/middleman-core/features/javascript-testing.feature

**Comment by jsoref** [comment]
@sandstrom, presumably you'd be reverting a significant portion of 3935a7c (at the very least deleting all of the javascript-app directory).

**Comment by sandstrom** [comment]
This is an old issue + I'm not the maintainer, so I don't decide anything 😄 

But my guess is that with the external pipeline and Middleman making the whole JS-thing pluggable we could probably drop all JS-tests from this library (including PhantomJS, which is sort of dead now).

**Comment by tdreyno** [comment]
Agreed. Thanks.
